src/hotspot/share/runtime/thread.cpp

```diff
@@ -2144,9 +2147,15 @@ void JavaThread::print_name_on_error(outputStream* st, char *buf, int buflen) co
 // JavaThread::print() is that we can't grab lock or allocate memory.
 void JavaThread::print_on_error(outputStream* st, char *buf, int buflen) const {
   st->print("%s \"%s\"", type_name(), get_thread_name_string(buf, buflen));
-  oop thread_obj = threadObj();
-  if (thread_obj != NULL) {
-    if (java_lang_Thread::is_daemon(thread_obj)) st->print(" daemon");
+  Thread* current = Thread::current_or_null();
+  if (current != nullptr && (!current->is_Java_thread() || JavaThread::cast(current)->is_oop_safe())) {
```

Comment:
Just realized, we can't have a null current thread if we are calling this as that is checked at a higher level. But we should be using `current_or_null_safe()` here as we could be in a signal-handling context.

Reply:
I used `Thread::current_or_null()` just to make the code more bullet proof. In all of
my testing I never ran into a case where `Thread::current()` returned `nullptr`.
I'm going to switch back to `Thread::current()` and remove the extra handling for
the `current == nullptr` case.

Reply:
If the code can be invoked in the context of a signal handler, as can happen for the error reporting path, then you must use `Thread::current_or_null_safe()` or risk introducing a deadlock or secondary crash.
It occurs to me that this may mean the existing safety-check is not correct because it may end up being checked in that context too.